module/zfs/dbuf.c

```diff
@@ -1764,6 +1764,96 @@ dbuf_unoverride(dbuf_dirty_record_t *dr)
 }
 
 /*
+ * Disassociate the frontend for any older transaction groups of a
+ * dbuf that is inside a range being freed.  The primary purpose is to
+ * ensure that the state of any dirty records affected by the operation
+ * remain consistent.
+ */
+static void
+dbuf_free_range_disassociate_frontend(dmu_buf_impl_t *db, dnode_t *dn,
+    dmu_tx_t *tx)
+{
+	uint64_t txg = tx->tx_txg;
+	dbuf_dirty_record_t *dr = list_head(&db->db_dirty_records);
+
+	ASSERT(dr != NULL);
+
+	if (dr->dr_txg == txg) {
+		/*
+		 * This buffer is "in-use", re-adjust the file
+		 * size to reflect that this buffer may
+		 * contain new data when we sync.
+		 */
+		if (db->db_blkid != DMU_SPILL_BLKID &&
+		    db->db_blkid > dn->dn_maxblkid)
+			dn->dn_maxblkid = db->db_blkid;
+		dbuf_unoverride(dr);
+	} else {
+		/*
+		 * This dbuf is not dirty in the open context.
+		 * Either uncache it (if its not referenced in
+		 * the open context) or reset its contents to
+		 * empty.
+		 */
+		dbuf_fix_old_data(db, txg);
+	}
+}
+
+
+static boolean_t
+dbuf_free_range_already_freed(dmu_buf_impl_t *db)
+{
+	/* XXX add comment about why these are OK */
+	if (db->db_state == DB_UNCACHED || db->db_state == DB_NOFILL ||
+	    db->db_state == DB_EVICTING) {
+		ASSERT(db->db.db_data == NULL);
+		mutex_exit(&db->db_mtx);
```

Comment:
We should not be dropping the mutex in a different function than where it was originally taken.  This is at least commented, but my personal feeling in these changes make this logic more complicated.  I'd much rather we not add these helper functions (dbuf_free_range_already_freed, dbuf_free_range_filler_will_free, dbuf_clear_successful) and instead add a block comment explaining each case below.